pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1119 @@
+				swap.ensure_currencies_match(foreign_swap, true)?;
+
+				match swap.amount.cmp(&foreign_swap.amount) {
+					Ordering::Equal => {
+						Ok(Self::SwapIntoForeignDone { done_swap: swap })
+					}
+					Ordering::Less => {
+						Ok(Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+							swap: Swap {
+								amount: foreign_swap.amount.ensure_sub(swap.amount)?,
+								..swap
+							},
+							done_amount: swap.amount,
+						})
+					},
+					Ordering::Greater => Err(DispatchError::Arithmetic(ArithmeticError::Overflow)),				}
+			},
+			// Increment done_foreign by swapped amount, leave invest amount untouched
+			Self::ActiveSwapIntoForeignCurrencyAndInvestmentOngoing {
+				swap: foreign_swap,
+				invest_amount,
+			} => {
+				swap.ensure_currencies_match(foreign_swap, true)?;
+
+				match swap.amount.cmp(&foreign_swap.amount) {
+					Ordering::Equal => {
+						Ok(Self::SwapIntoForeignDoneAndInvestmentOngoing {
+							done_swap: swap,
+							invest_amount: *invest_amount,
+						})
+					}
+					Ordering::Less => {
+						Ok(
+							Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+								swap: Swap {
+									amount: foreign_swap.amount.ensure_sub(swap.amount)?,
+									..swap
+								},
+								done_amount: swap.amount,
+								invest_amount: *invest_amount,
+							},
+						)
+					},
+					Ordering::Greater => Err(DispatchError::Arithmetic(ArithmeticError::Overflow)),
+				}
+			},
+			// Increment done_foreign by swapped amount
+			Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+				swap: foreign_swap,
+				done_amount,
+			} => {
+				swap.ensure_currencies_match(foreign_swap, true)?;
+				let done_amount = done_amount.ensure_add(swap.amount)?;
+
+				match swap.amount.cmp(&foreign_swap.amount) {
+					Ordering::Equal => {
+						Ok(Self::SwapIntoForeignDone {
+							done_swap: Swap {
+								amount: done_amount,
+								..swap
+							},
+						})
+					}
+					Ordering::Less => {
+						Ok(Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+							swap: Swap {
+								amount: foreign_swap.amount.ensure_sub(swap.amount)?,
+								..swap
+							},
+							done_amount,
+						})
+					},
+					Ordering::Greater => Err(DispatchError::Arithmetic(ArithmeticError::Overflow)),
+				}
+			},
+			// Increment done_foreign by swapped amount, leave invest amount untouched
+			Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+				swap: foreign_swap,
+				done_amount,
+				invest_amount,
+			} => {
+				swap.ensure_currencies_match(foreign_swap, true)?;
+				let done_amount = done_amount.ensure_add(swap.amount)?;
+
+				match swap.amount.cmp(&foreign_swap.amount) {
+					Ordering::Equal => {
+						Ok(Self::SwapIntoForeignDoneAndInvestmentOngoing {
+							done_swap: Swap {
+								amount: done_amount,
+								..swap
+							},
+							invest_amount: *invest_amount,
+						})
+					}
```

Comment:
Wanted to double-check this transition.

I do not understand why the `foreign_swap.amount` information is lost in the transition. I mean, no matter the `foreigh_swap.amount`, the `done_swap` resulting will be the same.

Reply:
This breaks me the "The sum of amounts must not change." invariant for this transition

Reply:
> This breaks the "The sum of amounts must not change." invariant for this transition

It should not. Does this happen in your tests? If so, could you share the lines with me?

We always bump `done_amount` by `swap.amount` and
* If the `swap.amount == foreign_swap.amount`, then we can simply remove `foreign_swap` from the state
* If `swap.amount < foreign_swap.amount`, we decrease `foreign_swap.amount` by `swap.amount`
* Else we throw

So we always remove `foreign_swap.amount` by `swap.amount` and increase `done_amount` by `swap_amount`, zero sum game.

Reply:
> This breaks the "The sum of amounts must not change." invariant for this transition

It should not. Does this happen in your tests? If so, could you share the lines with me?

We always bump `done_amount` by `swap.amount` and
* If the `swap.amount == foreign_swap.amount`, then we can simply remove `foreign_swap` from the state
* If `swap.amount < foreign_swap.amount`, we decrease `foreign_swap.amount` by `swap.amount`
* Else we throw

So we always remove `foreign_swap.amount` by `swap.amount` and increase `done_amount` by `swap_amount`, zero sum game.

Reply:
> This breaks the "The sum of amounts must not change." invariant for this transition

It should not. Does this happen in your tests? If so, could you share the lines with me?

We always bump `done_amount` by `swap.amount` and
* If the `swap.amount == foreign_swap.amount`, then we can simply remove `foreign_swap` from the state
* If `swap.amount < foreign_swap.amount`, we decrease `foreign_swap.amount` by `swap.amount`
* Else we throw

So we always remove `foreign_swap.amount` by `swap.amount` and increase `done_amount` by `swap_amount`, zero sum game.

Reply:
Yes, it comes from the test, I think the issue is when `swap.amount > foreign_swap.amount`. I did not rebase with your last change where you change that. Actually, I commented on your updated changes that currently are right haha. Then, the issue should be solved 👍🏻 

Thanks!